### DISCUSSION TITLE
Problem while building srsRAN from source #146

### DISCUSSION THREAD

**Discussion by wsq2002** [srsran/srsRAN_Project#719] [srsRAN-matlab] [answered]
Good morning, I encounter errors while building the make in the build directory.
I am on ubuntu 20.04.6. Someone had a similar error.

The result from the cmake output :
-- The C compiler identification is GNU 9.4.0
-- The CXX compiler identification is GNU 9.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build type not specified: defaulting to Release.
-- Performing Test HAS_MAYBE_UNINITIALIZED
-- Performing Test HAS_MAYBE_UNINITIALIZED - Success
-- Performing Test HAS_STRINGOP_OVERFLOW
-- Performing Test HAS_STRINGOP_OVERFLOW - Success
-- Performing Test HAVE_NON_VIRTUAL_DTOR
-- Performing Test HAVE_NON_VIRTUAL_DTOR - Success
-- Performing Test HAVE_SUGGEST_OVERRIDE
-- Performing Test HAVE_SUGGEST_OVERRIDE - Success
-- Performing Test HAVE_SHADOW
-- Performing Test HAVE_SHADOW - Success
-- Assertion level set to NORMAL
-- Could NOT find libdw (missing: LIBDW_LIBRARY LIBDW_INCLUDE_DIR)
-- Could NOT find libbfd (missing: LIBBFD_LIBRARY LIBBFD_INCLUDE_DIR)
-- Could NOT find libdwarf (missing: LIBDWARF_LIBRARY LIBDWARF_INCLUDE_DIR LIBELF_LIBRARY LIBELF_INCLUDE_DIR)
-- Found Backward: /home/a509/Desktop/srsRAN_Project/cmake/modules
-- Backward-cpp found, but external libraries are missing.
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1")
-- Checking for module 'mbedtls'
--   No package 'mbedtls' found
-- MBEDTLS LIBRARIES: /usr/lib/x86_64-linux-gnu/libmbedcrypto.so
-- MBEDTLS STATIC LIBRARIES: /usr/lib/x86_64-linux-gnu/libmbedcrypto.a
-- MBEDTLS INCLUDE DIRS: /usr/include
-- Found MbedTLS: /usr/lib/x86_64-linux-gnu/libmbedcrypto.so
-- Checking for module 'fftw3f >= 3.0'
--   Found fftw3f, version 3.3.7
-- FFTW3F LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.so
-- FFTW3F STATIC LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.a
-- FFTW3F INCLUDE DIRS: /usr/include
-- Found FFTW3F: /usr/lib/x86_64-linux-gnu/libfftw3f.so
-- Found GTest: /usr/lib/libgtest.a
-- Found Threads: TRUE
-- UHD LIBRARIES /usr/lib/x86_64-linux-gnu/libuhd.so
-- UHD INCLUDE DIRS /usr/include
-- Found UHD: /usr/lib/x86_64-linux-gnu/libuhd.so
-- Checking for module 'yaml-cpp'
--   Found yaml-cpp, version 0.5.2
-- YAMLCPP LIBRARIES: /usr/lib/x86_64-linux-gnu/libyaml-cpp.so
-- YAMLCPP INCLUDE DIRS: /usr/include
-- Found YAMLCPP: /usr/lib/x86_64-linux-gnu/libyaml-cpp.so
-- Performing Test HAVE_SSE
-- Performing Test HAVE_SSE - Success
-- SSE4.1 is enabled - target CPU must support it
-- Performing Test HAVE_AVX
-- Performing Test HAVE_AVX - Success
-- AVX is enabled - target CPU must support it
-- Performing Test HAVE_AVX2
-- Performing Test HAVE_AVX2 - Success
-- AVX2 is enabled - target CPU must support it
-- Performing Test HAVE_FMA
-- Performing Test HAVE_FMA - Success
-- FMA is enabled - target CPU must support it
-- Performing Test HAVE_AVX512
-- Performing Test HAVE_AVX512 - Failed
-- Performing Test HAVE_MARCH
-- Performing Test HAVE_MARCH - Success
-- Enabling ccache for C
-- Enabling ccache for CXX
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
-- Checking for module 'sctp'
--   No package 'sctp' found
-- SCTP LIBRARIES: /usr/lib/x86_64-linux-gnu/libsctp.so
-- SCTP INCLUDE DIRS: /usr/include
-- Found SCTP: /usr/lib/x86_64-linux-gnu/libsctp.so
-- FFT_LIBRARIES: /usr/lib/x86_64-linux-gnu/libfftw3f.so
-- Building srsRAN version 24.04.0
-- Configuring done (1.6s)
-- Generating done (0.8s)
-- Build files have been written to: /home/a509/Desktop/srsRAN_Project/build

I have all of the dependencies installed (I have double-checked)

My errors while building the "make -j $(nproc)" are the following :
[ 14%] Building CXX object lib/radio/uhd/CMakeFiles/srsran_radio_uhd.dir/radio_config_uhd_validator.cpp.o
In file included from /home/a509/Desktop/srsRAN_Project/lib/radio/uhd/radio_config_uhd_validator.h:25,
                 from /home/a509/Desktop/srsRAN_Project/lib/radio/uhd/radio_config_uhd_validator.cpp:23:
/home/a509/Desktop/srsRAN_Project/lib/radio/uhd/radio_uhd_device.h: In lambda function:
/home/a509/Desktop/srsRAN_Project/lib/radio/uhd/radio_uhd_device.h:231:39: error: ‘class uhd::usrp::multi_usrp’ has no member named ‘get_master_clock_rate_range’; did you mean ‘get_master_clock_rate’?
  231 |       uhd::meta_range_t range = usrp->get_master_clock_rate_range();
      |                                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                       get_master_clock_rate
make[2]: *** [lib/radio/uhd/CMakeFiles/srsran_radio_uhd.dir/build.make:76: lib/radio/uhd/CMakeFiles/srsran_radio_uhd.dir/radio_config_uhd_validator.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:10980: lib/radio/uhd/CMakeFiles/srsran_radio_uhd.dir/all] Error 2
make: *** [Makefile:146: all] Error 2

Wishing you a nice day.

**Comment by HaoxinSEU** [comment, marked as answer]
Hi,

Not sure if it's a problem. But how do you install libraries for UHD? If you directly install by `apt install`, it could be an outdated version. Try to add PPA from Ettus, and download the latest version.

Hope this helps.

**Reply by wsq2002** [reply]
Thank you for your help!
